File: src/tools/pmd-cpd/eval-repos/synthetic/typescript/heavy_duplication.ts
/**
 * TypeScript file with heavy duplication - multiple large duplicated blocks.
 */

interface ReportItem {
    id?: string;
    name?: string;
    email?: string;
    status?: string;
    createdAt?: string;
}

function generateUserReport(users: ReportItem[]): string {
    const lines: string[] = [];
    lines.push("=".repeat(60));
    lines.push("USER REPORT");
    lines.push("=".repeat(60));
    lines.push("");

    for (const item of users) {
        lines.push(`ID: ${item.id || "N/A"}`);
        lines.push(`Name: ${item.name || "Unknown"}`);
        lines.push(`Email: ${item.email || "N/A"}`);
        lines.push(`Status: ${item.status || "active"}`);
        lines.push(`Created: ${item.createdAt || "Unknown"}`);
        lines.push("-".repeat(40));
    }

    lines.push("");
    lines.push(`Total records: ${users.length}`);
    lines.push("=".repeat(60));
    return lines.join("\n");
}

function generateAdminReport(admins: ReportItem[]): string {
    const lines: string[] = [];
    lines.push("=".repeat(60));
    lines.push("ADMIN REPORT");
    lines.push("=".repeat(60));
    lines.push("");

    for (const item of admins) {
        lines.push(`ID: ${item.id || "N/A"}`);
        lines.push(`Name: ${item.name || "Unknown"}`);
        lines.push(`Email: ${item.email || "N/A"}`);
        lines.push(`Status: ${item.status || "active"}`);
        lines.push(`Created: ${item.createdAt || "Unknown"}`);
        lines.push("-".repeat(40));
    }

    lines.push("");
    lines.push(`Total records: ${admins.length}`);
    lines.push("=".repeat(60));
    return lines.join("\n");
}

function generateGuestReport(guests: ReportItem[]): string {
    const lines: string[] = [];
    lines.push("=".repeat(60));
    lines.push("GUEST REPORT");
    lines.push("=".repeat(60));
    lines.push("");

    for (const item of guests) {
        lines.push(`ID: ${item.id || "N/A"}`);
        lines.push(`Name: ${item.name || "Unknown"}`);
        lines.push(`Email: ${item.email || "N/A"}`);
        lines.push(`Status: ${item.status || "active"}`);
        lines.push(`Created: ${item.createdAt || "Unknown"}`);
        lines.push("-".repeat(40));
    }

    lines.push("");
    lines.push(`Total records: ${guests.length}`);
    lines.push("=".repeat(60));
    return lines.join("\n");
}

interface InputData {
    name?: string;
    email?: string;
    password?: string;
    age?: number;
}

function validateUserInput(data: InputData): string[] {
    const errors: string[] = [];
    if (!data.name) errors.push("Name is required");
    if (!data.email) errors.push("Email is required");
    if (!data.email?.includes("@")) errors.push("Invalid email format");
    if (!data.password) errors.push("Password is required");
    if ((data.password || "").length < 8) errors.push("Password must be at least 8 characters");
    if (!data.age) errors.push("Age is required");
    if ((data.age || 0) < 18) errors.push("Must be at least 18 years old");
    return errors;
}

function validateAdminInput(data: InputData): string[] {
    const errors: string[] = [];
    if (!data.name) errors.push("Name is required");
    if (!data.email) errors.push("Email is required");
    if (!data.email?.includes("@")) errors.push("Invalid email format");
    if (!data.password) errors.push("Password is required");
    if ((data.password || "").length < 8) errors.push("Password must be at least 8 characters");
    if (!data.age) errors.push("Age is required");
    if ((data.age || 0) < 18) errors.push("Must be at least 18 years old");
    return errors;
}

export { generateUserReport, generateAdminReport, generateGuestReport, validateUserInput, validateAdminInput };

File: src/tools/pmd-cpd/eval-repos/synthetic/csharp/HeavyDuplication.cs
/// <summary>
/// C# file with heavy duplication - multiple large duplicated blocks.
/// </summary>

using System;
using System.Collections.Generic;
using System.Text;

namespace Synthetic.HeavyDuplication
{
    public class ReportItem
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string Email { get; set; }
        public string Status { get; set; }
        public string CreatedAt { get; set; }
    }

    public static class ReportGenerator
    {
        public static string GenerateUserReport(List<ReportItem> users)
        {
            var lines = new List<string>();
            lines.Add(new string('=', 60));
            lines.Add("USER REPORT");
            lines.Add(new string('=', 60));
            lines.Add("");

            foreach (var item in users)
            {
                lines.Add($"ID: {item.Id ?? "N/A"}");
                lines.Add($"Name: {item.Name ?? "Unknown"}");
                lines.Add($"Email: {item.Email ?? "N/A"}");
                lines.Add($"Status: {item.Status ?? "active"}");
                lines.Add($"Created: {item.CreatedAt ?? "Unknown"}");
                lines.Add(new string('-', 40));
            }

            lines.Add("");
            lines.Add($"Total records: {users.Count}");
            lines.Add(new string('=', 60));
            return string.Join("\n", lines);
        }

        public static string GenerateAdminReport(List<ReportItem> admins)
        {
            var lines = new List<string>();
            lines.Add(new string('=', 60));
            lines.Add("ADMIN REPORT");
            lines.Add(new string('=', 60));
            lines.Add("");

            foreach (var item in admins)
            {
                lines.Add($"ID: {item.Id ?? "N/A"}");
                lines.Add($"Name: {item.Name ?? "Unknown"}");
                lines.Add($"Email: {item.Email ?? "N/A"}");
                lines.Add($"Status: {item.Status ?? "active"}");
                lines.Add($"Created: {item.CreatedAt ?? "Unknown"}");
                lines.Add(new string('-', 40));
            }

            lines.Add("");
            lines.Add($"Total records: {admins.Count}");
            lines.Add(new string('=', 60));
            return string.Join("\n", lines);
        }

        public static string GenerateGuestReport(List<ReportItem> guests)
        {
            var lines = new List<string>();
            lines.Add(new string('=', 60));
            lines.Add("GUEST REPORT");
            lines.Add(new string('=', 60));
            lines.Add("");

            foreach (var item in guests)
            {
                lines.Add($"ID: {item.Id ?? "N/A"}");
                lines.Add($"Name: {item.Name ?? "Unknown"}");
                lines.Add($"Email: {item.Email ?? "N/A"}");
                lines.Add($"Status: {item.Status ?? "active"}");
                lines.Add($"Created: {item.CreatedAt ?? "Unknown"}");
                lines.Add(new string('-', 40));
            }

            lines.Add("");
            lines.Add($"Total records: {guests.Count}");
            lines.Add(new string('=', 60));
            return string.Join("\n", lines);
        }
    }

    public class InputData
    {
        public string Name { get; set; }
        public string Email { get; set; }
        public string Password { get; set; }
        public int? Age { get; set; }
    }

    public static class Validator
    {
        public static List<string> ValidateUserInput(InputData data)
        {
            var errors = new List<string>();
            if (string.IsNullOrEmpty(data.Name)) errors.Add("Name is required");
            if (string.IsNullOrEmpty(data.Email)) errors.Add("Email is required");
            if (data.Email?.Contains("@") != true) errors.Add("Invalid email format");
            if (string.IsNullOrEmpty(data.Password)) errors.Add("Password is required");
            if ((data.Password ?? "").Length < 8) errors.Add("Password must be at least 8 characters");
            if (!data.Age.HasValue) errors.Add("Age is required");
            if ((data.Age ?? 0) < 18) errors.Add("Must be at least 18 years old");
            return errors;
        }

        public static List<string> ValidateAdminInput(InputData data)
        {
            var errors = new List<string>();
            if (string.IsNullOrEmpty(data.Name)) errors.Add("Name is required");
            if (string.IsNullOrEmpty(data.Email)) errors.Add("Email is required");
            if (data.Email?.Contains("@") != true) errors.Add("Invalid email format");
            if (string.IsNullOrEmpty(data.Password)) errors.Add("Password is required");
            if ((data.Password ?? "").Length < 8) errors.Add("Password must be at least 8 characters");
            if (!data.Age.HasValue) errors.Add("Age is required");
            if ((data.Age ?? 0) < 18) errors.Add("Must be at least 18 years old");
            return errors;
        }
    }
}
